App/functions/index.js

```diff
@@ -244,783 +245,748 @@ function emailTemplate(courseName,date,topics,results,type,quizCount,feedbackCou
         </body>
         </html>
         `
-      :
-      ''
-)
+                :
+                ''
+  )
 }
 async function getURLFromPasscode(passCode) {
   const db_ref = admin.app().database(url).ref('InternalDb/Courses/');
+  let snapshots;
   let courseURL;
-  await db_ref.orderByChild("passCode").equalTo(passCode).once("value",
-        function(snapshot) {
-          courseURL = Object.keys(snapshot.val())[0].replace(' ', '');
-        },
-        function(errorObject) {
-          console.log("The read failed: " + errorObject.code);
-        },
-  );
+  try {
+    snapshots = await db_ref.orderByChild("passCode").equalTo(passCode).once("value")
+    snapshots.forEach((snapshot) =>  {
```

Comment:
Any reason for not awaiting here ?

Reply:
A datasnapshot stores data about all its children in memory, so calling forEach on a datasnapshot isn't an asynchronous network operation. This can be verified [here](https://stackoverflow.com/questions/16098786/does-datasnapshot-object-contains-all-of-its-child-data)